src/Orleans.Runtime/Core/Dispatcher.cs

```diff
@@ -719,99 +349,107 @@ internal void SendResponse(Message request, Response response)
+                    MessagingStatisticsGroup.OnRejectedMessage(msg);
+                    Message response = this.messageFactory.CreateRejectionResponse(msg, Message.RejectionTypes.Unrecoverable,
+                        string.Format("SystemTarget {0} not active on this silo. Msg={1}", msg.TargetGrain, msg));
+                    this.messageCenter.SendMessage(response);
+                    this.logger.LogWarning(
+                        (int)ErrorCode.MessagingMessageFromUnknownActivation,
+                        "Received a message {Message} for an unknown SystemTarget: {Target}",
+                        msg,
+                        msg.TargetAddress);
+                    return;
                 }
-#endif
-                activation.ResetRunning(message);
 
-                // ensure inactive callbacks get run even with transactions disabled
-                if (!activation.IsCurrentlyExecuting)
-                    activation.RunOnInactive();
-
-                // Run message pump to see if there is a new request arrived to be processed
-                RunMessagePump(activation);
+                target.ReceiveMessage(msg);
             }
-        }
-
-        internal void RunMessagePump(ActivationData activation)
-        {
-            // Note: this method must be called while holding lock (activation)
-#if DEBUG
-            // This is a hot code path, so using #if to remove diags from Release version
-            // Note: Caller already holds lock on activation
-            if (logger.IsEnabled(LogLevel.Trace))
+            else if (messageCenter.TryDeliverToProxy(msg))
             {
-                logger.Trace(ErrorCode.Dispatcher_ActivationEndedTurn_Waiting,
-                    "RunMessagePump {0}: Activation={1}", activation.ActivationId, activation.DumpStatus());
+                return;
             }
-#endif
-            // don't run any messages if activation is not ready or deactivating
-            if (activation.State != ActivationState.Valid) return;
-
-            bool runLoop;
-            do
+            else
             {
-                runLoop = false;
-                var nextMessage = activation.PeekNextWaitingMessage();
-                if (nextMessage == null) continue;
-                if (!ActivationMayAcceptRequest(activation, nextMessage)) continue;
-                
-                activation.DequeueNextWaitingMessage();
-                // we might be over-writing an already running read only request.
-                HandleIncomingRequest(nextMessage, activation);
-                runLoop = true;
+                try
+                {
+                    var targetActivation = catalog.GetOrCreateActivation(
+                        msg.TargetAddress,
+                        msg.IsNewPlacement,
+                        msg.RequestContextData);
+
+                    if (targetActivation is null)
+                    {
+                        logger.LogInformation(
```

Comment:
Should we move this log line inside the `else` to avoid logging two times for the response case?

Reply:
Yes, well spotted, thanks